### PR TITLE
[Core] `aaz`: Support paginating calculation

### PR DESCRIPTION
**Pull request by necusjz**
**Related command**

**Description**

image

**Testing Guide**

`test_aaz_paging_with_token_and_limit`

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Core

File: src/azure-cli-core/azure/cli/core/aaz/_paging.py [hunk 1]
```diff
@@ -2,40 +2,79 @@
...
 
     def __next__(self):
+        def next_token(_, result):  # pylint: disable=unused-argument
+            from knack.log import get_logger
+            logger = get_logger(__name__)
+            logger.warning("Token of next page: %s", base64.b64encode(token).decode("utf-8"))
+
+        if self._total is not None and self._total < 0:
+            from knack.events import EVENT_CLI_SUCCESSFUL_EXECUTE
+            token = {"next_link": self._curr_link, "offset": self._page_size + self._total}
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_paging.py:35]: I think the offset should be `-self._total`

File: src/azure-cli-core/azure/cli/core/aaz/_paging.py [hunk 2]
```diff
@@ -2,40 +2,79 @@
...
         self._extract_result = extract_result
-        self._next_link = None
-        self._did_once_already = False
+        self._cli_ctx = cli_ctx
+        self._next_link = next_link
+        self._did_once_already = bool(next_link)  # desired start
+        self._total = page_size + offset if has_value(page_size) else None
+        self._start = offset
+        self._curr_link = None
+        self._page_size = None
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_paging.py:22]: recommend using another name to distinguish from  `page_size` argument

File: src/azure-cli-core/azure/cli/core/aaz/_paging.py [hunk 3]
```diff
@@ -2,40 +2,79 @@
...
     def __next__(self):
+        def next_token(_, result):  # pylint: disable=unused-argument
+            from knack.log import get_logger
+            logger = get_logger(__name__)
+            logger.warning("Token of next page: %s", base64.b64encode(token).decode("utf-8"))
+
+        if self._total is not None and self._total < 0:
+            from knack.events import EVENT_CLI_SUCCESSFUL_EXECUTE
+            token = {"next_link": self._curr_link, "offset": self._page_size + self._total}
+            token = json.dumps(token).encode("utf-8")
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_paging.py:36]: why not move these two lines in next_token func?

File: src/azure-cli-core/azure/cli/core/aaz/_paging.py [hunk 4]
```diff
@@ -2,40 +2,79 @@
...
+
+        if self._total < 0:
+            return iter(curr_page[start:self._total])  # smaller than current page
+
+        return iter(curr_page[start:])
 
 
-class AAZPaged(Iterator):
+class AAZPaged:
+    def __init__(self, executor, extract_result, cli_ctx, next_token=AAZUndefined, page_size=AAZUndefined):
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/aaz/_paging.py:64]: AAZUndefined is used for argument value. So it's not suitable to use it in page logic. Here, you should use None

File: src/azure-cli-core/azure/cli/core/tests/test_aaz_paging.py [hunk 5]
```diff
@@ -27,5 +28,39 @@ def executor(next_link):
...
+
+        def extract_result():
+            return result["value"], result["next_link"]
+
+        next_token = '{"next_link": 1, "offset": 1}'
+        paged = AAZPaged(
+            executor=executor, extract_result=extract_result, cli_ctx=DummyCli(), next_token=next_token, page_size=3
+        )
+        self.assertTrue(list(paged) == ["e", "f", "g"])
+
```
Comment (kairu-ms) [inline src/azure-cli-core/azure/cli/core/tests/test_aaz_paging.py:61]: please verify the token as well

**Review by kairu-ms — DISMISSED** [review]

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**